libtextsecure/message_receiver.js

```diff
@@ -984,45 +981,39 @@ MessageReceiver.prototype.extend({
 
     let conversation;
     try {
-      conversation = window.ConversationController.get(envelope.source);
+      conversation = await window.ConversationController.getOrCreateAndWait(envelope.source, 'private');
     } catch (e) {
       window.log.info('Error getting conversation: ', envelope.source);
     }
-
-    if (
-      envelope.preKeyBundleMessage &&
-      envelope.preKeyBundleMessage.type ===
-      textsecure.protobuf.PreKeyBundleMessage.Type.FRIEND_REQUEST
-    ) {
-      return this.handleFriendRequestMessage(envelope, content.dataMessage);
-    } else if (
-      envelope.type === textsecure.protobuf.Envelope.Type.CIPHERTEXT ||
-      // We also need to check for PREKEY_BUNDLE aswell if the session hasn't started.
-      // ref: libsignal-protocol.js:36120
-      envelope.type === textsecure.protobuf.Envelope.Type.PREKEY_BUNDLE
-    ) {
-      // If we get a cipher and we're already friends
-      // then we set our key exchange to complete
-      if (conversation && conversation.isFriend()) {
-        await conversation.setKeyExchangeCompleted();
-      }
+    if (envelope.type === textsecure.protobuf.Envelope.Type.FRIEND_REQUEST) {
+      conversation.onFriendRequestReceived();
     }
 
-    if (content.syncMessage) {
+    if (content.preKeyBundleMessage) {
+      const preKeyBundleMessage =
+        this.decodePreKeyBundleMessage(content.preKeyBundleMessage);
+      await this.savePreKeyBundleMessage(
+        envelope.source,
+        preKeyBundleMessage
+      );
+      return this.handleDataMessage(envelope, content.dataMessage, 'friend-request');
```

Comment:
we can leave this for now, but soon we'll need to handle prekey bundles for session resets, which should not have the 'friend-request' type passed in. So leave this for and I will make handleDataMessage figure out the type based off the envelope type etc.

Reply:
Ah true, so we will need a session reset envelope type?

Reply:
I don't think it will be necessary.
It will be easier if I tackle the session reset separately :)

Reply:
I'll leave the ball in your court